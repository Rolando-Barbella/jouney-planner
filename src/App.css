.App {
  height: 100vh; 
  display: flex; 
  align-items: center; 
  justify-content: center;
  flex-direction: column;
}

h1, h2 {
  font-weight: normal;
}

ul {
  padding: 0;
  list-style-type: none;
}

li {
  width: 3.7em;
  height: 3.7em;
  text-align: center;
  line-height: 3.7em;
  border-radius: 1.9em;
  margin-bottom: 1.3em;
  color: white;
  position: relative;
  background: #bbddff;
  border: solid 1px #8abdf0;
}

.journey-planner-container { 
  height: 70vh; 
  display: flex; 
  justify-content: center;
  flex-direction: column;
  padding: 12px;
  width: 50%;
  box-shadow: 1px 6px 1px 0 rgba(0,0,0,.1);
  border-radius: 4px;
  border: solid 1px #f6f6f6;
}

.journey-planner-container h2 {
  font-weight: 400;
  text-align: center;
}

.wrapper-journey-planner {
  display: flex;
  justify-content: space-between;
  margin-left: 20px;
  margin-right: 20px;
}

.result-journey {
  display: inline-block;
  margin-right: 10px;
}

.wrapper-result-journey {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
}

.wrapper-result-journey li::before{
  content: '';
  position: absolute;
  top: 1.65em;
  left: -1em;
  width: 5.5em;
  height: .2em;
  background: dodgerblue;
  z-index: -1;
}

.active-circle {
  background: dodgerblue;
  border: solid 1px #1873cc;
  font-weight: 500;
}

.button-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

button{
  background-image: linear-gradient(to top left,#4ca64c,#008000);
  color: #fff !important;
  display: inline-block;
  font-family: inherit;
  font-weight: 700;
  border: 0;
  border-radius: 2rem;
  white-space: nowrap;
  padding: 1rem 1.5rem;
  line-height: 1.4;
  position: relative;
  text-align: center;
  -webkit-transition: 70ms;
  transition: 70ms;
  position: relative;
  cursor: pointer;
  font-size: 14px;
}

button:hover{
  background-image: linear-gradient(to top left,#008000, #4ca64c );
}
