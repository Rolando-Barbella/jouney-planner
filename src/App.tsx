import React, { useState, useEffect } from 'react';
import Journey  from './components/Journey';
import { shortestJourney } from "./utils/shortest-journey";
import './App.css';

const fromAtoH: { [key: string]: number | object } = {
  start: { A: 1 },
  A: { C: 2 },
  C: { D: 1 },
  D: { F: 1 },
  F: { G: 3 },
  G: { H: 4 },
  H: { finish: 1 },
  finish: {}
};

const fromBtoH: { [key: string]: number | object } = {
  start: { B: 1 },
  A: { C: 4, B: 2 },
  B: { E: 7, D: 4 },
  C: { D: 6, },
  D: { G: 2, C: 4 },
  E: { H: 10 },
  G: { H: 4 },
  F: { G: 2 },
  H: { finish: 1 },
  finish: {}
};

const fromCtoH: { [key: string]: number | object } = {
  start: { C: 1 },
  C: { D: 1, F: 4 },
  D: { F: 1 },
  F: { G: 3 },
  G: { H: 4 },
  H: { finish: 1 },
  finish: {}
};

const fromDtoH: { [key: string]: number | object } = {
  start: { D: 1 },
  D: { F: 1, G: 2 },
  F: { G: 3 },
  G: { H: 4 },
  H: { finish: 1 },
  finish: {}
};

function App() {
  const [ startJourneis ] = useState(['A', 'B', 'C', 'D']);
  const [ endJourneis ]= useState(['H']);
  const [ journey, setJourney ] = useState({ start: '', end: 'H' });
  const [ showJourneyPlanner, setShowJourneyPlanner ] = useState(true);
  
  const disabledButton = Object.values(journey).every(j => j !== "");

  const findShortestJourney = (journey:{[ key: string ]: string } ) => {
    let { start } = journey;
    if (start === "A") {
      return shortestJourney(fromAtoH);
    }

    if (start === "B") {
      return shortestJourney(fromBtoH);
    }

    if (start === "D") {
      return shortestJourney(fromDtoH);
    }

    if (start === "C") {
      return shortestJourney(fromCtoH);
    }
    return { distance: 0, path: [] };
  }

  const { path } = findShortestJourney(journey);

  return (
    <div className="App">
      <header>
        <h1>Journey Planner</h1>
      </header>
      <section className="journey-planner-container">
        {showJourneyPlanner ?
          <>
            <div className="wrapper-journey-planner">
              <div>
                <h2>Start</h2>
                <Journey 
                  dataTestId="start-options"
                  journeis={startJourneis}
                  journeyName={journey.start}
                  onClick={(start: string) => setJourney({ ...journey, start })} 
                />
              </div>
              <div className="end-wrapper">
                <br/>
                <Journey 
                  dataTestId="end-options"
                  journeis={endJourneis}
                  journeyName={journey.end}
                />
              </div>
            </div>
            <div className="button-container">
              <button
                data-testid="search-journey"
                disabled={!disabledButton}
                style={{ background: !disabledButton? 'grey': "" }}
                onClick={() => { 
                  setShowJourneyPlanner(false);
                  findShortestJourney(journey);
                }}
              >
                Search
              </button>
            </div>
          </> : (
            <div className="wrapper-result-journey">
              <Journey 
                dataTestId="result-journey"
                verticalRow="result-journey"
                journeis={path}
              />
              <div className="button-container">
                <button 
                  data-testid="reset-journey"
                  onClick={() => { 
                    setShowJourneyPlanner(true);
                    setJourney({ start: '', end: 'H' });
                  }}
                >
                  Reset
                </button>
              </div>
            </div>
          )
        }
      </section>
    </div>
  );
}

export default App;
