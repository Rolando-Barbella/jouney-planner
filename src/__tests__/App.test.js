import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import App from '../App';

// afterEach(cleanup);

it('renders without crashing', () => {
  const { asFragment } = render(<App />);

  expect(asFragment()).toMatchSnapshot();
});

it('should display start title', () => {
  let startTitle = 'Start';
  const { getByText } = render(
    <App>{startTitle}</App>
  );
  
  expect(getByText(startTitle)).toBeInTheDocument();
});

it('should display end title', () => {
  const endTitle = 'End';
  const { getByText } = render(
    <App>{endTitle}</App>
  );

  expect(getByText(endTitle)).toBeInTheDocument()
});

it('it displays start options', () => {
  const { getByTestId } = render(<App />);
  const startOptions = getByTestId('start-options');

  expect(startOptions.children.length).toBe(4);
});

it('displays end options', () => {
  const { getByTestId } = render(<App/>);
  const endOptions = getByTestId('end-options');

  expect(endOptions.children.length).toBe(4);
});

it('displays result journey from B to H', () => {
  const { getByTestId, getByText } = render(<App/>);
  
  fireEvent.click(getByText('B'));
  fireEvent.click(getByText('H'));
  fireEvent.click(getByTestId('search-journey'));
  const resultJourney = getByTestId('result-journey');
  
  expect(resultJourney.children.length).toBe(4);
});
