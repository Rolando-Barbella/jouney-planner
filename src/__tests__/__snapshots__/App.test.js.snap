// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <h1>
        Journey Planner
      </h1>
    </header>
    <section
      class="journey-planner-container"
    >
      <div
        class="wrapper-journey-planner"
      >
        <div>
          <h2>
            Start
          </h2>
          <ul
            data-testid="start-options"
          >
            <li
              class=""
            >
              A
            </li>
            <li
              class=""
            >
              B
            </li>
            <li
              class=""
            >
              C
            </li>
            <li
              class=""
            >
              D
            </li>
          </ul>
        </div>
        <div
          class="yo"
        >
          <br />
          <ul
            data-testid="end-options"
          >
            <li
              class="active-circle"
            >
              H
            </li>
          </ul>
        </div>
      </div>
      <div
        class="button-container"
      >
        <button
          data-testid="search-journey"
          disabled=""
          style="background: grey;"
        >
          Search
        </button>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
